src/Features/CSharp/Portable/MakeFieldReadonly/CSharpMakeFieldReadonlyDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,42 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.MakeFieldReadonly;
+
+namespace Microsoft.CodeAnalysis.CSharp.MakeFieldReadonly
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    internal class CSharpMakeFieldReadonlyDiagnosticAnalyzer :
+        AbstractMakeFieldReadonlyDiagnosticAnalyzer<IdentifierNameSyntax, ConstructorDeclarationSyntax>
+    {
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxNodeAction(AnalyzeType, SyntaxKind.ClassDeclaration, SyntaxKind.StructDeclaration);
+
+        protected override ISyntaxFactsService GetSyntaxFactsService()
+            => CSharpSyntaxFactsService.Instance;
+
+        protected override bool IsWrittenTo(IdentifierNameSyntax name, SemanticModel model, CancellationToken cancellationToken)
+            => name.IsWrittenTo();
+
+        protected override bool IsMemberOfThisInstance(SyntaxNode node)
```

Comment:
This is would not needed to be language specific if written as an IOperation analyzer.

Reply:
➡️ Resolved by filing #25652

Reply:
I tried to write this as an `IOperation` analzyer before. It was a lot cleaner, but failed because the given API's didn't provide something about property setters, I believe. It should be in some of the previous review comments, but some of the discussion probably happened in gitter.

Reply:
Yes, detecting property writes is still not trivial, and @333fred had a tracking issue to improve the IOperation APIs for it. However, given this analyzer only deals with fields, we should consider making it an IOperation analyzer, field writes should be easier to detect.

---
In reply to: [176229105](https://github.com/dotnet/roslyn/pull/19067#discussion_r176229105) [](ancestors = 176229105)